plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,648 @@
+#define FU_SYNAPTICS_CAPE_MODULE_RC_CMD_MODE_UNSUPPORTED   (-1056)
+
+/* CAPE HID message structure. Little endian.*/
+typedef struct __attribute__((packed)) {
+	gint16 data_len : 16; /*number of dwords*/
+	guint16 cmd_id : 15;  /*command id*/
+	guint16 reply : 1;    /*Host want a reply from device, 1 = true*/
+	guint32 module_id;    /*module id*/
+	guint32 data[FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN]; /*Command data*/
+} FuCapCmd;
+
+/* CAPE HID report structure.*/
+typedef struct __attribute__((packed)) {
+	guint16 report_id; /*two bytes of report id, this should be 1*/
+	FuCapCmd cmd;
+} FuCapCmdHidReport;
+
+/* CAPE Commands */
+typedef enum {
+	FU_SYNAPTICS_CMD_FW_UPDATE_START = 0xC8,	  /* notifies firmware update started*/
+	FU_SYNAPTICS_CMD_FW_UPDATE_WRITE = 0xC9,	  /* updates firmware data */
+	FU_SYNAPTICS_CMD_FW_UPDATE_END = 0xCA,		  /* notifies firmware update finished*/
+	FU_SYNAPTICS_CMD_MCU_SOFT_RESET = 0xAF,		  /* reset device*/
+	FU_SYNAPTICS_CMD_FW_GET_ACTIVE_PARTITION = 0x1CF, /* gets cur active partition number*/
+	FU_SYNAPTICS_CMD_GET_VERSION = 0x103,		  /* gets cur firmware version*/
+} FuCommand;
+
+/*GET flag*/
+#define FU_SYNAPTICS_CMD_GET_FLAG 0x100
+
+/* CAPE Fuupd device structure*/
+struct _FuSynapticsCapeDevice {
+	FuHidDevice parent_instance;
+	guint32 active_partition; /*active partition, 0 or 1*/
+	gchar *version;
+};
+
+G_DEFINE_TYPE(FuSynapticsCapeDevice, fu_synaptics_cape_device, FU_TYPE_HID_DEVICE)
+
+static gboolean
+fu_synaptics_cape_device_sendcmd_ex(FuSynapticsCapeDevice *self,
+				    FuCapCmd *req,
+				    gulong delay_us,
+				    GError **error);
+
+static gboolean
+fu_synaptics_cape_device_sendcmd(FuSynapticsCapeDevice *self,
+				 const guint32 module_id,
+				 const guint32 cmd_id,
+				 const guint32 *data,
+				 const guint32 data_len,
+				 const gulong delay_us,
+				 GError **error);
+
+static void
+fu_synaptics_cape_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+	fu_common_string_append_ku(str, idt, "Active Partition", self->active_partition);
+	fu_common_string_append_kv(str, idt, "FW version", self->version);
+}
+
+typedef enum {
+	FU_SYNAPTICS_CXAUDIO_OPERATION_READ,
+	FU_SYNAPTICS_CXAUDIO_OPERATION_WRITE,
+	FU_SYNAPTICS_CXAUDIO_OPERATION_LAST
+} FuSynapticsCapeOperation;
+
+typedef enum {
+	FU_SYNAPTICS_CXAUDIO_OPERATION_FLAG_NONE = 0,
+	FU_SYNAPTICS_CXAUDIO_OPERATION_FLAG_VERIFY = (1 << 4),
+} FuSynapticsCapeOperationFlags;
+
+/* reset device */
+static gboolean
+fu_synaptics_cape_device_reset(FuSynapticsCapeDevice *self, GError **error)
+{
+	gboolean result;
+
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GTimer) timer = g_timer_new();
+	g_return_val_if_fail(FU_IS_SYNAPTICS_CAPE_DEVICE(self), FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (!(result =
+		  fu_synaptics_cape_device_sendcmd(self,
+						   FU_SYNAPTICS_CAPE_CMD_APP_ID('C', 'T', 'R', 'L'),
```

Comment:
I'm not sure what result is doing here

Reply:
oops, I forgot to remove this unused variable.